hw/mcu/microchip/pic32mz2048efg100/src/hal_timer.c

```diff
@@ -0,0 +1,548 @@
+            break;
+        case 5:
+            IPC8CLR = _IPC8_T7IP_MASK | _IPC8_T7IS_MASK;
+            IPC8SET = 3 << _IPC8_T7IP_POSITION;
+            IFS1CLR = _IFS1_T7IF_MASK;
+            IEC1SET = _IEC1_T7IE_MASK;
+            break;
+        case 6:
+            IPC9CLR = _IPC9_T8IP_MASK | _IPC9_T8IS_MASK;
+            IPC9SET = 3 << _IPC9_T8IP_POSITION;
+            IFS1CLR = _IFS1_T8IF_MASK;
+            IEC1SET = _IEC1_T8IE_MASK;
+            break;
+        case 7:
+            IPC10CLR = _IPC10_T9IP_MASK | _IPC10_T9IS_MASK;
+            IPC10SET = 3 << _IPC10_T9IP_POSITION;
+            IFS1CLR = _IFS1_T9IF_MASK;
+            IEC1SET = _IEC1_T9IE_MASK;
+            break;
+    }
+}
+
+static void
+hal_timer_disable_int(int timer_num)
+{
+    switch (timer_num) {
+        case 0:
+            IEC0CLR = _IEC0_T2IE_MASK;
+            IFS0CLR = _IFS0_T2IF_MASK;
+            break;
+        case 1:
+            IEC0CLR = _IEC0_T3IE_MASK;
+            IFS0CLR = _IFS0_T3IF_MASK;
+            break;
+        case 2:
+            IEC0CLR = _IEC0_T4IE_MASK;
+            IFS0CLR = _IFS0_T4IF_MASK;
+            break;
+        case 3:
+            IEC0CLR = _IEC0_T5IE_MASK;
+            IFS0CLR = _IFS0_T5IF_MASK;
+            break;
+        case 4:
+            IEC0CLR = _IEC0_T6IE_MASK;
+            IFS0CLR = _IFS0_T6IF_MASK;
+            break;
+        case 5:
+            IEC1CLR = _IEC1_T7IE_MASK;
+            IFS1CLR = _IFS1_T7IF_MASK;
+            break;
+        case 6:
+            IEC1CLR = _IEC1_T8IE_MASK;
+            IFS1CLR = _IFS1_T8IF_MASK;
+            break;
+        case 7:
+            IEC1CLR = _IEC1_T9IE_MASK;
+            IFS1CLR = _IFS1_T9IF_MASK;
+            break;
+    }
+
+}
+
+static void
+compute_next_expiry(struct pic32_timer *bsp_timer)
+{
+    struct hal_timer *first = TAILQ_FIRST(&bsp_timer->hal_timer_queue);
+    if (first != NULL) {
+        bsp_timer->expiry = first->expiry;
+        bsp_timer->expiry *= bsp_timer->interrupt_per_tick;
+    }
+}
+
+static void
+handle_timer_list(int timer_num)
+{
+    int i;
+    int timers_to_remove = 0;
+    uint32_t current_tick = hal_timer_read(timer_num);
+    struct hal_timer *entry;
+
+    TAILQ_FOREACH(entry, &timers[timer_num].hal_timer_queue, link) {
+        if (entry->expiry <= current_tick) {
+            entry->cb_func(entry->cb_arg);
+            ++timers_to_remove;
+        } else {
+            break;
+        }
+    }
+
+    for (i = 0; i < timers_to_remove; ++i) {
+        TAILQ_REMOVE(&timers[timer_num].hal_timer_queue,
+                     TAILQ_FIRST(&timers[timer_num].hal_timer_queue), link);
+    }
+    if (timers_to_remove != 0) {
+        compute_next_expiry(&timers[timer_num]);
+    }
+}
+
+void
+__attribute__((interrupt(IPL3AUTO), vector(_TIMER_2_VECTOR)))
+timer2_isr(void)
+{
+    ++timers[0].interrupt_count;
+
+    if (!(TAILQ_EMPTY(&timers[0].hal_timer_queue)) &&
+        timers[0].expiry <= timers[0].interrupt_count) {
```

Comment:
Yes, hal_timer is independent of os tick. These are meant to be used to do things when user needs to do things with a finer granularity of timing than what OS ticker provides. E.g. low-layer radio stuff, bitbanging output etc.

Yes for the 2nd statement as well. It is better to take less number of interrupts! If you have compare register available, then you could use those. Also note that you might need to deal with overflows, if your TMRx/PRx are 32-bit registers. Hardware most likely already provides you an interrupt source for that. If there is no such capability, you can use compare register for that same purpose.

Hope this helps!